util/pull/pull.go

```diff
@@ -17,17 +17,18 @@ import (
 	"github.com/moby/buildkit/util/imageutil"
 	"github.com/moby/buildkit/util/progress/logs"
 	"github.com/moby/buildkit/util/pull/pullprogress"
-	"github.com/moby/buildkit/util/resolver"
 	"github.com/moby/buildkit/util/resolver/limited"
 	"github.com/moby/buildkit/util/resolver/retryhandler"
 	digest "github.com/opencontainers/go-digest"
 	ocispecs "github.com/opencontainers/image-spec/specs-go/v1"
 	"github.com/pkg/errors"
 )
 
+type SessionResolver func(g session.Group) remotes.Resolver
+
 type Puller struct {
 	ContentStore content.Store
-	Resolver     *resolver.Resolver
+	Resolver     remotes.Resolver
```

Comment:
Make this `RegistryResolver` and add another `OCIResolver`. Trying to map the into same type just adds complexity. This also allows cleaning up `PullManifests`

Reply:
I'm looking at this again and this is already containerd resolver so I'm not sure what needs these `WithImageStore` functions. If you can clean up these functions in `ocilayout.go` and get rid of this `SessionResolver` callback then this can remain `remotes.Resolver`.

Reply:
I don't understand what you are asking to do here, or how changes would clean `PullManifests`.

This is a generic `Puller` that, given a `remotes.Resolver`, can use the funcs there (`Resolve`, `Fetcher` and from there `Fetch`) to retrieve blobs. I don't think there is anything left in here that is remote registry vs OCI layout specific anymore?

Reply:
Oh, I missed your comment a few mins ahead of mine.

I think I got it. Take a look.